Input/output JSON file for courses

raynertjx opened the pull request:
Add courses json file to data directory
    
Course data is currently not being stored in a human
editable file.

We need to store course data, such as lessons, in a
human editable file.

Created a courses.json file in the data directory for
users to modify.

To ensure there is a human editable file for users to
modify course data.

Closes #56

raynertjx committed:
Add courses json file to data directory

Course data is currently not being stored in a human
editable file.

We need to store course data, such as lessons, in a
human editable file.

Created a courses.json file in the data directory for
users to modify.

To ensure there is a human editable file for users to
modify course data.

raynertjx committed:
Add resetData method to Courses

This is to reset the course data that is present in the
courses.json file.

raynertjx committed:
Create courses.json file on execution of command

To help initialise a courses.json file if one does
not exist, upon running a command.

raynertjx committed:
Fill courses.json file with sample data

Courses.json file is blank when created.

It needs to be populated with sample data for users
to work with.

Let's,
* write to the courses.json file with sample data when necessary

lshaoqin commented on src/main/java/seedu/address/model/Courses.java:
```diff
@@ -0,0 +1,78 @@
+     */
+    public void resetData(ReadOnlyCourses newData) {
+        requireNonNull(newData);
+        setCourses(CourseData.getCourseList());
```
This should still use newData and not grab sample data by default.

lshaoqin commented on src/main/java/seedu/address/model/course/UniqueCourseList.java:
```diff
@@ -1,23 +1,25 @@
  * Represents a list of courses with TAs in SOC. Immutable.
  */
 public class UniqueCourseList implements Iterable<Course> {
     private static final ObservableList<Course> internalList = FXCollections.observableArrayList(
```
This list should not be static under the new implementation. The main inconvenience would be that findByCourseCode would need to be drilled in order to be used.

lshaoqin commented on src/main/java/seedu/address/model/course/UniqueCourseList.java:
```diff
@@ -1,23 +1,25 @@
-            CourseData.getCourseList()
-    );
+            SampleDataUtil.getSampleCourses());
     private static final ObservableList<Course> internalUnmodifiableList =
```
We can consider getting rid of this list if it is not used.

lshaoqin commented on src/main/java/seedu/address/model/course/UniqueCourseList.java:
```diff
@@ -47,6 +49,34 @@ public static Lesson findLesson(String courseCode, String name) {
+        internalList.add(toAdd);
+    }
+
+    public void setCourses(List<Course> courses) {
```
internalList being non-static should fix the issues with this method

lshaoqin reviewed (changes requested):
Great effort, the code quality seems good otherwise but it seems to be bugged when a json file is present :(

I've proposed some changes to fix this bug.

raynertjx committed:
Remove Courses and ReadOnlyCourses classes

Courses and ReadOnlyCourses classes are redundant in the
parsing of the courses.json file.

A simpler approach would be more than ideal to handle the
parsing of the file.

Let's,
* use the static list in the UniqueCourseList and keep it
as the source of truth

lshaoqin committed:
Correct language for logs

lshaoqin reviewed (approved):
Nice job!

lshaoqin committed:
Remove extra file

lshaoqin committed:
Fix bug with find after altering course json
